File: cs/Problems/Day16.cs
namespace AOC2023.Problems;

public class Day16 : IProblem<int>
{
    public int Solve(ReadOnlySpan<char> input) => CountEnergizedLightTiles(input);

    private int CountEnergizedLightTiles(ReadOnlySpan<char> input)
    {
        Span<Range> lineRanges = stackalloc Range[110];
        int height = input.Split(lineRanges, Environment.NewLine);
        int width = input[lineRanges[0]].Length;

        var matrix = new char[height, width];
        var history = new Direction[height, width];

        for (int j = 0; j < height; j++)
        for (int i = 0; i < width; i++)
            matrix[j, i] = input[lineRanges[j]][i];

        var beam = new FacingPoint(0, 0, Direction.Right);
        StartLightBeam(beam, matrix, history);

        int tileCount = 0;

        for (int j = 0; j < height; j++)
        for (int i = 0; i < width; i++)
            tileCount += history[j, i] > 0 ? 1 : 0;

        return tileCount;
    }

    private static bool StartLightBeam(FacingPoint beam, char[,] matrix, Direction[,] history)
    {
        int width = matrix.GetLength(1);
        int height = matrix.GetLength(0);

        while (!beam.IsOffset(width, height))
        {
            var instruction = matrix[beam.Y, beam.X];
            var visits = history[beam.Y, beam.X];

            if (visits.HasFlag(beam.Direction))
                break;

            history[beam.Y, beam.X] = visits | beam.Direction;

            _ = (instruction, beam.Direction) switch
            {
                ('/', Direction.Right) => beam.Move(Direction.Up),
                ('/', Direction.Left) => beam.Move(Direction.Down),
                ('/', Direction.Up) => beam.Move(Direction.Right),
                ('/', Direction.Down) => beam.Move(Direction.Left),
                ('\\', Direction.Right) => beam.Move(Direction.Down),
                ('\\', Direction.Left) => beam.Move(Direction.Up),
                ('\\', Direction.Up) => beam.Move(Direction.Left),
                ('\\', Direction.Down) => beam.Move(Direction.Right),
                ('|', Direction.Right) => StartLightBeam(beam.Split(Direction.Up), matrix, history),
                ('|', Direction.Left) => StartLightBeam(beam.Split(Direction.Up), matrix, history),
                ('-', Direction.Up) => StartLightBeam(beam.Split(Direction.Right), matrix, history),
                ('-', Direction.Down) => StartLightBeam(beam.Split(Direction.Right), matrix, history),
                _ => beam.Move(beam.Direction)
            };
        }

        return true;
    }

    record struct FacingPoint(int X, int Y, Direction Direction)
    {
        public bool Move(Direction direction)
        {
            if (direction == Direction.Up)
                Y--;
            else if (direction == Direction.Down)
                Y++;
            else if (direction == Direction.Right)
                X++;
            else if (direction == Direction.Left)
                X--;

            Direction = direction;
            return true;
        }

        public FacingPoint Split(Direction direction)
        {
            var split = this with { };
            split.Move(direction);

            var oppositeDirection = direction switch
            {
                Direction.Right => Direction.Left,
                Direction.Left => Direction.Right,
                Direction.Up => Direction.Down,
                Direction.Down => Direction.Up,
                _ => direction
            };
            Move(oppositeDirection);

            return split;
        }

        public readonly bool IsOffset(int width, int height) => X < 0 || Y < 0 || X >= width || Y >= height;
    }

    [Flags]
    enum Direction : byte
    {
        Right = 1,
        Left = 2,
        Up = 4,
        Down = 8
    };
}

File: ts/problems/day16.ts
import { NEW_LINE } from "../inputReader.js";

const countEnergizedLightTiles = (input: string) => startBeamAndCountTiles(input.split(NEW_LINE));

const startBeamAndCountTiles = (matrix: string[]) => {
  const history: number[][] = Array.from(matrix).map(() => new Array(matrix[0].length).fill(0));
  const beam: FacingPoint = { x: 0, y: 0, direction: Direction.Right };

  startLightBeam(beam, matrix, history);

  // By this point history will be complete. Count all items > 0.
  return history.flat().reduce((tileCount, cell) => tileCount + Number(cell > 0), 0);
};

const startLightBeam = (beam: FacingPoint, matrix: string[], history: number[][]) => {
  const height = matrix.length;
  const width = matrix[0].length;

  while (!isOffsetBeam(beam, width, height)) {
    const visits = history[beam.y][beam.x];
    if (visits & beam.direction) break;

    history[beam.y][beam.x] |= beam.direction;

    const instruction = matrix[beam.y][beam.x];
    const navigation = `${instruction},${beam.direction}`;

    if (navigation === "|,1" || navigation === "|,2") {
      startLightBeam(splitLightBeam(beam, Direction.Up), matrix, history);
    } else if (navigation === "-,4" || navigation === "-,8") {
      startLightBeam(splitLightBeam(beam, Direction.Right), matrix, history);
    } else if (navigation in beamInstructionMap) {
      beamInstructionMap[navigation](beam);
    } else {
      beamInstructionMap.wildcard(beam);
    }
  }
};

const moveLightBeam = (lightBeam: FacingPoint, direction: Direction) => navigationMap[direction](lightBeam);

const splitLightBeam = (lightBeam: FacingPoint, direction: Direction) => {
  const newBeam = moveLightBeam({ ...lightBeam }, direction);
  moveLightBeam(lightBeam, oppositeDirectionMap[direction]);
  return newBeam;
};

const isOffsetBeam = (beam: FacingPoint, width: number, height: number) =>
  beam.x < 0 || beam.y < 0 || beam.x >= width || beam.y >= height;

enum Direction {
  Right = 1,
  Left = 2,
  Up = 4,
  Down = 8,
}

const navigationMap: Record<Direction, (point: FacingPoint) => FacingPoint> = {
  [Direction.Up]: (point: FacingPoint) => (point.y--, (point.direction = Direction.Up), point),
  [Direction.Down]: (point: FacingPoint) => (point.y++, (point.direction = Direction.Down), point),
  [Direction.Left]: (point: FacingPoint) => (point.x--, (point.direction = Direction.Left), point),
  [Direction.Right]: (point: FacingPoint) => (point.x++, (point.direction = Direction.Right), point),
};

const oppositeDirectionMap = {
  [Direction.Up]: Direction.Down,
  [Direction.Down]: Direction.Up,
  [Direction.Left]: Direction.Right,
  [Direction.Right]: Direction.Left,
};

const beamInstructionMap: Record<string, (beam: FacingPoint) => void> = {
  "/,1": (beam) => moveLightBeam(beam, Direction.Up),
  "/,2": (beam) => moveLightBeam(beam, Direction.Down),
  "/,4": (beam) => moveLightBeam(beam, Direction.Right),
  "/,8": (beam) => moveLightBeam(beam, Direction.Left),
  "\\,1": (beam) => moveLightBeam(beam, Direction.Down),
  "\\,2": (beam) => moveLightBeam(beam, Direction.Up),
  "\\,4": (beam) => moveLightBeam(beam, Direction.Left),
  "\\,8": (beam) => moveLightBeam(beam, Direction.Right),
  wildcard: (beam) => moveLightBeam(beam, beam.direction),
};

type FacingPoint = {
  x: number;
  y: number;
  direction: Direction;
};

export default countEnergizedLightTiles;
